html/beps/bep_0046.rst

```diff
@@ -0,0 +1,238 @@
+have been redundant had the peer not been upload-only, and send them if the peer
+changes state to no longer be upload-only. Such bulk update could be compact
+with the extension message proposed in the following section.
+
+.. _BEP21: http://bittorrent.org/beps/bep_0021.html
+
+Compressed bitfields
+====================
+
+This BEP proposes a new extension message which announces multiple pieces in a
+single message in a space-efficient manner. The content of the message is
+conceptually an entire bitfield with bits set for the new pieces the peer has
+received.
+
+Extension message name is "lt_have", as specified in the extension handshake.
+
+This message may substitute the BITFIELD message.
+
+This bitfield is not transferred verbatim, it is *run length encoded*, as a
+sequence of commands. Each command is encoded as a *byte aligned* block. The
+first two bits of each block indicates which command it encodes. That is, the
+most significant bits of the first byte.
+
+These are the commands:
+
+0. fill forward *n* bytes of zeroes (0x00)
+1. fill forward *n* bytes of ones (0xff)
+2. consume the next *n* bytes as verbatim bits of the bitfield
+3. fill forward *n* bytes of zeroes and consume one byte of verbatim bitfield
+
+The start position is always at the first piece in the torrent, position 0. All
+commands progress the cursor forward, towards higher piece indices.
+
+The ``num_bytes`` field in all command blocks are *unsigned* and *big-endian*.
+
+Since there is no point in ever encoding 0 bytes in the ``num_bytes`` field,
+the number of bytes following is ``num_bytes`` + 1. i.e. if the ``num_bytes``
+field is 0, it means 1 byte.
+
+If the message ends before reaching the last piece, the remaining bits are
+assumed to be zero.
+
+A block may run past the last piece position by *at most* 7 bits. This is
+required since all blocks are byte aligned. A run that extends past the last
+piece by more than 7 bits should be considered a malformed message.
+
+fill forward zeros
+..................
+
+::
+
+	0       8        16
+	+--+----+--------+
+	|00| num_bytes   |
+	|  | (14 bits)   |
+	+--+----+--------+
+
+The first 2 bits are 00, the next 14 bits encode the number of *bytes* of
+bitfield this zero-run is.
+
+fill forward ones
+.................
+
+::
+
+	0       8        16
+	+--+----+--------+
+	|01| num_bytes   |
+	|  | (14 bits)   |
+	+--+----+--------+
+
+The first two bits are 01, the next 14 bits encode the number of *bytes* of
+bitfield this one-run is.
+
+verbatim block
+..............
+
+.. parsed-literal::
+
+	0       8        16
+	+--+----+--------+----------- - -
+	\|10| num_bytes   | *variable length*
+	|  | (14 bits)   |
+	+--+----+--------+----------- - -
+
+The first two bits are 10, the next 14 bits are the number of *bytes* of the
+command block are to be interpreted as verbatim parts of the bitfield (at the
+current position).
+
+verbatim position
```

Comment:
This last command was basically what I came up with to assign some meaning to the last slot. A careful observer will notice that this message, despite paying an extra byte for the extension message type, comes out to the exact same size as a regular HAVE message (which uses 32 bits for a piece index). Granted, this can't reach quite as far as a 32 bit index, but I imagine the vast majority of torrents, even with 16 kiB pieces, would have fewer than 2^14 * 8 pieces.

The benefit of this is that with small pieces, it's likely that you still want to download them in clusters, to play nice with your disk cache, and hence, you can announce 8 pieces for the price of one regular HAVE. Anyway, this seems like the least valuable part of the RLE scheme